### ISSUE TITLE
Support language localization in static embedding

### ISSUE THREAD

**Issue by dtslvr** [metabase/metabase#8490]
Similar to https://github.com/metabase/metabase/issues/8384 I'd like to embed a dashboard in a defined language.

My proposal is that you can pass the locale like this:
```
var iframeUrl = METABASE_SITE_URL + "/embed/dashboard/" + token + "#bordered=true&titled=true&locale=de-DE";
```

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by flamber** [comment]
From 0.36.0 it will be possible to change the language with the header `X-Metabase-Locale` #12566

**Comment by godric** [comment]
👋 Is there any plan to support this or other options ?

We're trying to embed our dashboard via iframes and would like to be able to match our users's locales.
@dtslvr's proposition sounds exactly like what we need, while @flamber's solution wouldnt work since you can't set Headers on `<iframes>`.

**Comment by flamber** [comment]
@godric You can add headers via a reverse-proxy based on e.g. a URL or browser language headers.

**Comment by tkssharma** [comment]
@flamber  can you please explain this a little bit 
i am also facing same issue

**Comment by flamber** [comment]
@tkssharma I don't know which reverse-proxy you're using, but with Nginx you would do something like this:
```nginx
# Metabase: Use this to automatically set the language based on (likely) primary language of the browser
map $http_accept_language $metabase_browser_locale {
    default en;
    ~*^(\w\w) $1;
}

# Metabase: Use this to manually define the language via the URL argument ?locale=<lang> - e.g. https://metabase.example.com/public/dashboard/e64a258f-1234-4567-9dab-8daba24bffbc?locale=es
# Note - it has to be an argument, not part of the hash, since hashes are not sent to the server, so following will *not* work: https://metabase.example.com/public/dashboard/e64a258f-1234-4567-9dab-8daba24bffbc?parameter=test#locale=es
map $request_uri $metabase_url_locale {
    default en;
    ~*\??\&?locale=(\w\w) $1;
}

server {
    server_name metabase.example.com;
    return 301 https://metabase.example.com$request_uri;
}

server {
    server_name metabase.example.com;

    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    ssl_certificate /etc/letsencrypt/live/metabase.example.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/metabase.example.com/privkey.pem;
    ssl_trusted_certificate /etc/letsencrypt/live/metabase.example.com/fullchain.pem;

    location ~ ^/\.well-known/acme-challenge { root /var/www/_letsencrypt; allow all; }

    location / {
        proxy_http_version 1.1;
        proxy_read_timeout 600;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

# Metabase: Uncomment one of the following proxy_set_header below
#        proxy_set_header X-Metabase-Locale $metabase_browser_locale;
#        proxy_set_header X-Metabase-Locale $metabase_url_locale;

        proxy_pass http://127.0.0.1:3000/;
    }
}
```

**Comment by tkssharma** [comment]
@flamber  This may be another simple option 

it's just we need language based on what the UI client wants 
As you said we need a header that we can achieve while rendering the dashboard URL in Iframe 

1.  just to validate this use custom header addon on browsers add request header `X-Metabase-Locale` with value `en-US`
2. you can see the dashboard now coming in passed locale-code, irrespective of what you have set language at the account level 

- The proxy can be one solution but too much for just getting this working
- this should also be achieved by simple iframe rendering on UI side, iframe request will send UI locale-code 

```
<iframe id='myIframe' src=""></iframe> 

var myIframe = document.querySelector('#myIframe');
var myURL= '<iframe-url>'

populateIframe(myIframe, myURL, [['X-Metabase-Locale', 'en-US']]);

function populateIframe(iframe, url, headers) {
  var xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.onreadystatechange = handler;
  xhr.responseType = 'blob';
  headers.forEach(function(header) {
    xhr.setRequestHeader(header[0], header[1]);
  });
  xhr.send();

  function handler() {
    if (this.readyState === this.DONE) {
      if (this.status === 200) {
        iframe.src = URL.createObjectURL(this.response);
      } else {
        console.error('Request failed', this);
      }
    }
  }
}
```

**Comment by romulus-ai** [comment]
> @tkssharma I don't know which reverse-proxy you're using, but with Nginx you would do something like this:
> 
> ```nginx
> ### Metabase: Use this to automatically set the language based on (likely) primary language of the browser
> map $http_accept_language $metabase_browser_locale {
>     default en;
>     ~*^(\w\w) $1;
> }
> 
> ### Metabase: Use this to manually define the language via the URL argument ?locale=<lang> - e.g. https://metabase.example.com/public/dashboard/e64a258f-1234-4567-9dab-8daba24bffbc?locale=es
> ### Note - it has to be an argument, not part of the hash, since hashes are not sent to the server, so following will *not* work: https://metabase.example.com/public/dashboard/e64a258f-1234-4567-9dab-8daba24bffbc?parameter=test#locale=es
> map $request_uri $metabase_url_locale {
>     default en;
>     ~*\??\&?locale=(\w\w) $1;
> }
> 
> server {
>     server_name metabase.example.com;
>     return 301 https://metabase.example.com$request_uri;
> }
> 
> server {
>     server_name metabase.example.com;
> 
>     listen 443 ssl http2;
>     listen [::]:443 ssl http2;
>     ssl_certificate /etc/letsencrypt/live/metabase.example.com/fullchain.pem;
>     ssl_certificate_key /etc/letsencrypt/live/metabase.example.com/privkey.pem;
>     ssl_trusted_certificate /etc/letsencrypt/live/metabase.example.com/fullchain.pem;
> 
>     location ~ ^/\.well-known/acme-challenge { root /var/www/_letsencrypt; allow all; }
> 
>     location / {
>         proxy_http_version 1.1;
>         proxy_read_timeout 600;
>         proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> 
> ### Metabase: Uncomment one of the following proxy_set_header below
> #        proxy_set_header X-Metabase-Locale $metabase_browser_locale;
> #        proxy_set_header X-Metabase-Locale $metabase_url_locale;
> 
>         proxy_pass http://127.0.0.1:3000/;
>     }
> }
> ```

I used this example to implement the locale configuration via request URL, so er use "&locale=es" for example to configure spain.. However when we load a dashboard, only the first, the intial, request contains this "locale" parameter, after this request another bunch of requests are generated by metabase to load the graphs of the dashboard and non of them contains the locale parameter. This means the text contained in the graphs is in the default language but the control menu on top is i.e. in Spain.. Anyone has a solution for that?

**Comment by flamber** [comment]
@romulus-ai There's work being done currently to make it easier to set language for embeds #22561
Until then, you would have to do something where you store a session cookie, which you can then use in all following requests.

**Comment by paoliniluis** [comment]
This was included in v44 but only for full app embedding

**Comment by albertoperdomo** [comment]
This is supported since 44 in

* regular app usage, e.g. `https://mymb.metabaseapp.com/dashboard/1-a-look-at-invoices?locale=de-DE`
* interactive embedding, e.g. `https://mymb.metabaseapp.com/dashboard/1-a-look-at-invoices?locale=de-DE`
* static embedding, e.g. `    iframeUrl = urljoin(METABASE_SITE_URL, "/embed/dashboard/" + token + "?locale=de-DE#theme=night&background=false&bordered=true&titled=true")
* public links, e.g. `https://mymb.metabaseapp.com/public/dashboard/f7f32d9b-28f0-4b5b-b121-26c8f348a883?locale=de-DE`

It's not properly documented, so we will address that soon.

**Comment by albertoperdomo** [comment]
We've discovered that while the parameter is supported, it has unexpected effects which break downloads in public links and static embedding. Work is still needed to tackle this properly.

**Comment by WiNloSt** [comment]
Downloads with correct translation would be handled in https://github.com/metabase/metabase/issues/49531. Now the downloads are only partially translated.

**Comment by WiNloSt** [comment]
From my testing with @npretto we found that only static embed question failed to download the result when using with `?locale=xxx`, other combination didn't fail the download.

**Comment by rodrigoGA** [comment]
@WiNloSt I have updated to version v0.52.1, and the use of locale in embeddings has been lost.
The Metabase changelog states:
```
Language localization in static embedding with hash parameter #locale=xx
```
Currently, I have many static embeddings, and all dashboards are designed for English.
The dashboards were being embedded as follows:
```
...?locale=en#...
```
And the locale worked fine. Since the update, this functionality is broken.
I tried moving the locale to:
```
...#locale=en&...
```
But this does not work. Have they introduced a bug, or has a feature that users intentionally relied on been removed?
